test/Microsoft.ML.Core.Tests/UnitTests/TestUtilities.cs

```diff
@@ -0,0 +1,169 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.ML.Internal.Utilities;
+using Microsoft.ML.RunTests;
+using Microsoft.ML.Trainers.FastTree.Internal;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Microsoft.ML.Core.Tests.UnitTests
+{
+    public class TestUtilities : BaseTestBaseline
+    {
+        public TestUtilities(ITestOutputHelper helper)
+            : base(helper)
+        {
+        }
+
+        [Fact]
+        [TestCategory("Utilities")]
+        public void CheckIsSortedInt()
+        {
+            // A sorted (increasing) array
+            int[] x = Enumerable.Range(0, 10).ToArray();
+            Assert.True(Utils.IsSorted(x));
+
+            // A monotonically increasing array
+            var x1Temp = x[1];
+            var x7Temp = x[7];
+            x[1] = x[0];
+            x[7] = x[6];
+            Assert.True(Utils.IsSorted(x));
+            x[1] = x1Temp;
+            x[7] = x7Temp;
+
+            // Not sorted
+            x[1] = x[6];
+            Assert.False(Utils.IsSorted(x));
+            x[1] = x1Temp;
+        }
+
+        [Fact]
+        [TestCategory("Utilities")]
+        public void CheckIsSortedFloat()
+        {
+            // A sorted (increasing) array
+            List<float> x = Enumerable.Range(0, 1000000).Select(i => (float)i).ToList();
+            Assert.True(Utils.IsSorted(x));
+
+            // A monotonically increasing array
+            var x1Temp = x[1];
+            var x7Temp = x[7];
+            x[1] = x[0];
+            x[7] = x[6];
+            Assert.True(Utils.IsSorted(x));
+            x[1] = x1Temp;
+            x[7] = x7Temp;
+
+            // Not sorted
+            x[1] = x[6];
+            Assert.False(Utils.IsSorted(x));
+            x[1] = x1Temp;
+            
+            // NaN: `Array.Sort()` will put NaNs into the first position,
+            // but we want to guarantee that NaNs aren't allowed in these arrays.
+            var x0Temp = x[0];
+            x[0] = float.NaN;
+            Assert.False(Utils.IsSorted(x));
+            x[0] = x0Temp;
+        }
+
+        [Fact]
+        [TestCategory("Utilities")]
+        public void CheckIsIncreasing()
+        {
+            // An increasing array
+            int[] x = Enumerable.Range(0, 10).ToArray();
+            Assert.True(Utils.IsIncreasing(0, x, 10));
+            // Check the lower bound
+            Assert.False(Utils.IsIncreasing(1, x, 10));
+            // The upper bound should be exclusive
+            Assert.False(Utils.IsIncreasing(0, x, 9));
+            // Any len should work
+            Assert.True(Utils.IsIncreasing(0, x, 5, 10));
```

Comment:
What about if x contains duplicate numbers in a sorted order (like 0,1,1,2,3,)?

Reply:
Covered by the lines just below this one! (I refer to this as monotonically increasing sequences.)

Reply:
oh thats true, ok!

---
In reply to: [250402110](https://github.com/dotnet/machinelearning/pull/2214#discussion_r250402110) [](ancestors = 250402110)